pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/spf13/cobra"
+
+	"github.com/devfile/library/pkg/devfile"
+	"github.com/devfile/library/pkg/devfile/parser"
+
+	"github.com/redhat-developer/odo/pkg/catalog"
+	_init "github.com/redhat-developer/odo/pkg/init"
+	"github.com/redhat-developer/odo/pkg/init/registry"
+	"github.com/redhat-developer/odo/pkg/log"
+	"github.com/redhat-developer/odo/pkg/odo/cli/init/asker"
+	"github.com/redhat-developer/odo/pkg/odo/cli/init/params"
+	"github.com/redhat-developer/odo/pkg/odo/cmdline"
+	"github.com/redhat-developer/odo/pkg/odo/genericclioptions"
+	odoutil "github.com/redhat-developer/odo/pkg/odo/util"
+	"github.com/redhat-developer/odo/pkg/preference"
+	scontext "github.com/redhat-developer/odo/pkg/segment/context"
+	"github.com/redhat-developer/odo/pkg/testingutil/filesystem"
+
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+// RecommendedCommandName is the recommended command name
+const RecommendedCommandName = "init"
+
+var initExample = templates.Examples(`
+  # Boostrap a new component in interactive mode
+  %[1]s
+`)
+
+type InitOptions struct {
+	// Backends to build init parameters
+	backends []params.ParamsBuilder
+
+	// filesystem on which command is running
+	fsys filesystem.Filesystem
+
+	// Clients
+	initClient       _init.Client
+	preferenceClient preference.Client
+
+	// the parameters needed to run the init procedure
+	params.InitParams
+
+	// Destination directory
+	contextDir string
+}
+
+// NewInitOptions creates a new InitOptions instance
+func NewInitOptions(backends []params.ParamsBuilder, fsys filesystem.Filesystem, initClient _init.Client, prefClient preference.Client) *InitOptions {
+	return &InitOptions{
+		backends:         backends,
+		fsys:             fsys,
+		initClient:       initClient,
+		preferenceClient: prefClient,
+	}
+}
+
+// Complete will build the parameters for init, using different backends based on the flags set,
+// either by using flags or interactively if no flag is passed
+// Complete will return an error immediately if the current working directory is not empty
+func (o *InitOptions) Complete(cmdline cmdline.Cmdline, args []string) (err error) {
+
+	o.contextDir, err = o.fsys.Getwd()
+	if err != nil {
+		return err
+	}
+
+	empty, err := isEmpty(o.fsys, o.contextDir)
+	if err != nil {
+		return err
+	}
+	if !empty {
+		return errors.New("The current directory is not empty. You can bootstrap new component only in empty directory.\nIf you have existing code that you want to deploy use `odo deploy` or use `odo dev` command to quickly iterate on your component.")
+	}
+
+	flags := cmdline.GetFlags()
+	done := false
+	for _, backend := range o.backends {
+		if backend.IsAdequate(flags) {
+			o.InitParams, err = backend.ParamsBuild()
+			if err != nil {
+				return err
+			}
+			done = true
+			break
+		}
+	}
+
+	if !done {
+		odoutil.LogErrorAndExit(nil, "no backend found to build init parameters. This should not happen")
+	}
+
+	scontext.SetComponentType(cmdline.Context(), o.InitParams.Devfile)
```

Comment:
This does not work when using `--devfile-path` flag.

Reply:
I proposed an other way to do it